File: LeetCode/c/116.c
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  struct TreeLinkNode *left, *right, *next;
 * };
 *
 */

void connect(struct TreeLinkNode *root) {
    if (!root)
        return;
    Support(root, NULL);
}

void Support(struct TreeLinkNode *node, struct TreeLinkNode *next) {
    if (node) {
        if (node->left && node->right)
            node->left->next = node->right;
        if (node->right && next && next->left)
            node->right->next = next->left;
        Support(node->left, node->right);
        Support(node->right, next && next->left ? next->left : NULL);
    }
}

File: LeetCode/python/116.py
class Solution:
    # @param root, a tree link node
    # @return nothing
    def Support(self, node, next_node):
        if node:
            if node.left and node.right:
                node.left.next = node.right
            if node.right and next_node and next_node.left:
                node.right.next = next_node.left
            self.Support(node.left, node.right)
            self.Support(node.right, next_node.left if next_node and next_node.left else None)

    def connect(self, root):
        if not root:
            return
        self.Support(root, None)